### ISSUE TITLE
[Question] Is it necessary to have two ways to deal with kl divergence?

### ISSUE THREAD

**Issue by scris** [verl-project/verl#509]
As described in https://github.com/volcengine/verl/issues/455#issuecomment-2695996653:
1. When use_kl_loss is False，we add kl_penalty in ray_trainer.py, and metric is recored to critic/kl. https://github.com/volcengine/verl/blob/65cceb3c9dbe6e18230bc3dc045af6e3280c9752/verl/trainer/ppo/ray_trainer.py#L978-L984
2. When use_kl_loss is True, we add kl_penalty in worker/actor/dp_actor update_policy(), and metric is recorded to actor/kl_loss and actor/ppo_kl.
https://github.com/volcengine/verl/blob/65cceb3c9dbe6e18230bc3dc045af6e3280c9752/verl/workers/actor/dp_actor.py#L285-L305

Why will we need to have these two ways to deal with kl divergence, and one is suggested to be used with ppo, while the other is suggested to be used with grpo? Are there any particular (or historical) reasons? Many thanks!
https://github.com/volcengine/verl/blob/65cceb3c9dbe6e18230bc3dc045af6e3280c9752/verl/trainer/config/ppo_trainer.yaml#L33

Plus, I also want to know why both actor/kl_loss and actor/ppo_kl are recorded when use_kl_loss is True. Thanks again!

**Comment by asirgogogo** [comment]
i have a question!

why does the GRPO algorithm need to use entropy_loss? Do I need to set entropy_coeff to 0?

**Comment by wzq016** [comment]
One way is to use KL as a process reward, and one way is to use KL as a loss. You can regard them as a PRM vs. ORM here.

KL can be used as PRM because its log format can be decomposed into a summation sequence.

GRPO uses a different kl estimator that is incompatible with PRM and can be used as ORM only.

**Comment by scris** [comment]
@wzq016 Thanks for your sharing. However I do also find some replication experiment that uses use_kl_loss = False for GRPO, unlike suggested from veRL.

For example, under the TinyZero project that uses veRL:
- use_kl_loss = False
  - https://wandb.ai/jiayipan/TinyZero/runs/m19na0qi/overview
  - https://wandb.ai/jiayipan/TinyZero/runs/m19na0qi/workspace
  - Official wandb result from the author, Jiayi Pan, and the result seems to be alright
- use_kl_loss = True
  - https://github.com/Jiayi-Pan/TinyZero/issues/63
  - The result seems to be problematic

Do you have any ideas on that? Thank you very much!

**Comment by scris** [comment]
@asirgogogo Entropy loss is mainly designed for encouraging the model to enhance the diversity of responses, and it's actually a "bonus", not "loss" added. 
https://github.com/volcengine/verl/blob/65cceb3c9dbe6e18230bc3dc045af6e3280c9752/verl/workers/actor/dp_actor.py#L294
https://github.com/volcengine/verl/blob/65cceb3c9dbe6e18230bc3dc045af6e3280c9752/verl/workers/actor/dp_actor.py#L304
And someone (https://github.com/Open-Reasoner-Zero/Open-Reasoner-Zero/issues/4) has suggested to eliminate kl loss, and you may also suggest to eliminate entropy loss if you find the result is better.

**Comment by wzq016** [comment]
@scris Good question! KL prevents reward hacking by limiting parameters near the initial checkpoint.

When the reward is unlikely hackable, such as a rule-based reward used in TinyZero, you can drop KL for potentially better performance.

**Comment by gxy-gxy** [comment]
> [@asirgogogo](https://github.com/asirgogogo) Entropy loss is mainly designed for encouraging the model to enhance the diversity of responses, and it's actually a "bonus", not "loss" added in my opinion.
> 
> [verl/verl/workers/actor/dp_actor.py](https://github.com/volcengine/verl/blob/65cceb3c9dbe6e18230bc3dc045af6e3280c9752/verl/workers/actor/dp_actor.py#L294)
> 
> Line 294 in [65cceb3](/volcengine/verl/commit/65cceb3c9dbe6e18230bc3dc045af6e3280c9752)
> 
>  policy_loss = pg_loss - entropy_loss * entropy_coeff 
> 
> [verl/verl/workers/actor/dp_actor.py](https://github.com/volcengine/verl/blob/65cceb3c9dbe6e18230bc3dc045af6e3280c9752/verl/workers/actor/dp_actor.py#L304)
> 
> Line 304 in [65cceb3](/volcengine/verl/commit/65cceb3c9dbe6e18230bc3dc045af6e3280c9752)
> 
>  policy_loss = policy_loss + kl_loss * self.config.kl_loss_coef 
> 
> And as someone ([Open-Reasoner-Zero/Open-Reasoner-Zero#4](https://github.com/Open-Reasoner-Zero/Open-Reasoner-Zero/issues/4)) has suggested to eliminate kl loss, you may also suggest to eliminate entropy loss if you find the result is better.

Is there any reference paper or blog that shows the effectiveness of a negative entropy loss? Thanks!

**Comment by Kirrito-k423** [comment]
Thanks for the discussion. Based on the current code, it is useful to separate the two KL paths:

1. `algorithm.use_kl_in_reward` applies KL before advantage/policy updates by subtracting a KL penalty from token-level rewards. This is the reward-shaping path and is implemented in `verl/trainer/ppo/ray_trainer.py`.
2. `actor_rollout_ref.actor.use_kl_loss` applies KL inside the actor objective, after the policy loss has been computed. This is configured in `verl/trainer/config/actor/actor.yaml` and implemented in `verl/workers/utils/losses.py`.

For GRPO, the concern raised above is that adding token-level KL into rewards can interfere with the relative sequence-level reward/advantage estimate. That is why `use_kl_loss=True` is often recommended for GRPO-style training, while in-reward KL is closer to the PPO/RLHF reward-shaping setup. There are still tasks where disabling KL, or using a different KL strength, can work better, especially when the reward is hard to hack or strongly rule-based.

For the metrics: `actor/ppo_kl` is a monitoring metric from the PPO-style policy update, measuring the approximate KL between the current policy and the old policy used for the update. The KL loss path instead uses the reference policy and contributes to the objective through `kl_loss * kl_loss_coef`. So these metrics are related but not duplicates; they track different comparisons.

中文总结：当前代码里确实保留了两条 KL 路径。`use_kl_in_reward` 是把 KL 当作 reward penalty，在 reward/advantage 之前扣掉；`use_kl_loss` 是把 KL 当作 actor loss 的一部分，直接加到 policy loss 上。GRPO 场景下更推荐后者，主要是为了避免 token-level KL reward 干扰相对序列 reward 的 advantage 估计。`actor/ppo_kl` 只是 policy update 中当前策略相对 old policy 的监控指标，不等同于 reference-policy KL loss。
This response was generated by a scheduled Codex GPT5.5 issue processing task.
